networkx/algorithms/structuralholes.py

```diff
@@ -138,6 +138,55 @@ def redundancy(G, u, v, weight=None):
         )
         return 1 - r
 
+    try:
+        import numpy as np
+
+        has_numpy = True
+    except:
+        has_numpy = False
+
+    if nodes is None and has_numpy:
+        # In order to compute constraint of all nodes,
+        # algorithms based on sparse matrices can be much faster
+
+        # Obtain the adjacency matrix
+        P = nx.adjacency_matrix(G, weight=weight)
+
+        # Calculate mutual weights
+        mutual_weights1 = P + P.T
+        mutual_weights2 = mutual_weights1.copy()
+
+        # Mutual_weights1 = Normalize mutual weights by row sums
+        sum_mutual_weights = mutual_weights1.sum(axis=1)
+        val = np.asarray(
+            np.repeat(sum_mutual_weights, mutual_weights1.getnnz(axis=0))
+        ).flatten()
+        mutual_weights1.data = mutual_weights1.data / val
+
+        # Mutual_weights2 = Normalize mutual weights by row max
+        max_mutual_weights = mutual_weights2.max(axis=1).todense()
+        val = np.asarray(
+            np.repeat(max_mutual_weights, mutual_weights2.getnnz(axis=0))
+        ).flatten()
+        mutual_weights2.data = mutual_weights2.data / val
```

Comment:
Re: the implementation, I think this can be significantly simplified by just relying on array operations. I assume most of the code here is to avoid divide-by-zero warnings, but those can simply be ignored:

```suggestion
        with np.errstate(divide="ignore"):
            # Mutual_weights1 = Normalize mutual weights by row sums
            mutual_weights1 /= mutual_weights1.sum(axis=1)[:, np.newaxis]

            # Mutual_weights2 = Normalize mutual weights by row max
            mutual_weights2 /= mutual_weights2.max(axis=1).toarray()
```

Reply:
Many thanks to the suggestions!
It's true that many of these codes are designed to avoid the divide-by-zero warnings.